[PERF] account: Speed-up duplicate bills lookup in list view

pivi-odoo opened the pull request:
Description:
------------
Following commit 4e7b5e44b292dc41217e4c50048b8bb8408246f5, the list view for bills now shows a colored highlight on the bill's reference if there is a potential duplicate, and the matching on the invoice date was relaxed to include potential bills from the same year.

This caused a performance regression, as the index on `invoice_date` could no longer be used since the year is now extracted from it.

This patch adds a partial 'btree' index on the `ref` field specifically for bills, as it is the most discriminatory factor when identifying potential duplicates.

Benchmark:
----------
On a database with 6M `account_move` records, 230k of which are `'in_invoice'` & `'in_refund'`, `_fetch_duplicate_reference`, triggered when opening the list view of bills, took:

| Before | After | Speedup |
|--------|-------|---------|
| 19.8s  | 20ms  | 990x    |


---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

jco-odoo commented:
@RicGR98 @pivi-odoo I am not sure why this year thing or the invoice date is in the ref anyways.  I would think the ref should contain the entire identification, but I remember @tsb-odoo once saying that people mess around with it.  But also functionally, if you put an invoice the 31st of December instead of 1st of January, you won't see the duplicate.   To me, the index should above all have the commercial_partner_id, which it does not.  (but why the huge speedup then?)

RicGR98 commented:
> @RicGR98 @pivi-odoo I am not sure why this year thing or the invoice date is in the ref anyways. I would think the ref should contain the entire identification, but I remember @tsb-odoo once saying that people mess around with it. But also functionally, if you put an invoice the 31st of December instead of 1st of January, you won't see the duplicate. To me, the index should above all have the commercial_partner_id, which it does not. (but why the huge speedup then?)

The year thing is important because some reset their sequence every year, so there might be multiple bills with the same ref but on different years :)

I'm not a big big fan of escaping when one could have `"` and `'` but it's subjective so IMO the PR LGTM, thanks a lot, I hadn't thought about the perfs ! I'll let you double-check Josse :)

pivi-odoo commented:
> @RicGR98 @pivi-odoo I am not sure why this year thing or the invoice date is in the ref anyways. I would think the ref should contain the entire identification, but I remember @tsb-odoo once saying that people mess around with it. But also functionally, if you put an invoice the 31st of December instead of 1st of January, you won't see the duplicate. To me, the index should above all have the commercial_partner_id, which it does not. (but why the huge speedup then?)

@jco-odoo The speed up comes from the fact the plan changed from a non-selective Bitmap Index Scans on `company_id` AND `move_type` indexes, which churned through 40 GiB of data 🫠, to just an Index Scan on this new index, which is strongly selective (only 5.6 MiB of data touched). So it's an IO bottleneck mostly.
The most overfit index for this specific query would be based on `(ref, commercial_partner_id, company_id, move_type)`.
The order I'm showing is from most selective to least selective. This was inferred from Postgres' `pg_stats.n_distinct` catalog, `ref` has 2x more distinct values than `commercial_partner_id`, therefor is more selective. I also confirmed this by testing the same index with permutations of the keys of the index, it's fastest in the order I've mentioned.

Since the index was kinda fat, I just removed the keys that were non-selective, and keeping `commercial_partner_id` yielded no significant performance improvement (1ms vs 3ms execution time), here the planning takes the most time (16ms), so I dropped it. 

Here are the plans:
- Before: https://explain.dalibo.com/plan/2eee67ee5c12c85b
- After: https://explain.dalibo.com/plan/37a8872cg1552edf

> I'm not a big big fan of escaping when one could have " and ' but it's subjective so IMO the PR LGTM, thanks a lot, I hadn't thought about the perfs ! I'll let you double-check Josse :)

I heard you guys use `"..."` primarily for user-facing strings / translated messages, and `'...'` for technical strings. Since this is a technical string, I just did the necessary escaping.
I'm totally impartial on the usage `"..."` vs `'...'`. Let me know what you prefer.

jco-odoo commented on addons/account/models/account_move.py:
```diff
@@ -749,6 +749,13 @@ def init(self):
+            indexname='account_move_duplicate_bills_idx',
+            tablename='account_move',
+            expressions=['ref'],
+            where='move_type IN (\'in_invoice\', \'in_refund\')',
```
```suggestion
            where="move_type IN ('in_invoice', 'in_refund')",
```
That is less ugly to me...

jco-odoo commented:
@pivi-odoo Thanks for explaining.  Ref has a trigram search, but that is not ideal for exact matching, so a separate one with bigram just for duplicate checking makes sense.  Let me know for the style stuff.  Otherwise lgtm.
